This GIT repository contains all the code associated to the paper: Seawater Microbes in Reef Monitoring: Gene Content is a Strong Predictor of Water Chemistry Across the Great Barrier Reef

A read-based metagenomics analysis was applied to separate taxonomic and functional profiling of seawater microbiomes in offshore GBR reefs and elucidate the role of environmental filtering and functional redundancy in shaping reef bacterioplankton communities separately at taxonomic and functional levels (see Louca et al. 2016; Song et al. 2022).

Demultiplexed raw reads were first quality-checked in **FastQC** (version 0.11.3) (Andrews, 2010) and quality-filtered in **Trimmomatic** (version 0.38) (Bolger et al. 2014) to trim barcodes/adapters and remove low-quality bases (Phred <20). Quality-filtered reads were then aligned against the NCBI nr database using the **DIAMOND** (version 2.0.9) alignment tool (Buchfink et al. 2015). For each read, the top match reported by Diamond with e-value of <10-5 was retained to exclude poor annotations. Resulting Diamond files (in daa format) were then imported into **MEGAN** (version 6.23.0) (Huson et al. 2016) for community profiling. Raw microbial abundance counts were exported from MEGAN for genus-level taxonomic and functional (GO-terms) composition, and subsequently imported into **R Studio** (version 1.4.1717) (R Core Team, 2023)
